doc/sphinx/user-extensions/syntax-extensions.rst

```diff
@@ -7,64 +7,179 @@ In this chapter, we introduce advanced commands to modify the way Coq
 parses and prints objects, i.e. the translations between the concrete
 and internal representations of terms and commands.
 
-The main commands to provide custom symbolic notations for terms are
-:cmd:`Notation` and :cmd:`Infix`; they will be described in the
-:ref:`next section <Notations>`. There is also a
-variant of :cmd:`Notation` which does not modify the parser; this provides a
-form of :ref:`abbreviation <Abbreviations>`. It is
-sometimes expected that the same symbolic notation has different meanings in
-different contexts; to achieve this form of overloading, Coq offers a notion
-of :ref:`notation scopes <Scopes>`.
-The main command to provide custom notations for tactics is :cmd:`Tactic Notation`.
+When inputing a term to Coq, it goes through several successive steps.
+
+* :gdef:`lexing`: the input is split in a linear stream of tokens, for
+  instance the string ``"1 + 2 * 2"`` will be split into the sequence
+  of five tokens `'1' '+' '2' '*' '2'`
+
+* :gdef:`parsing`: the previous stream is given a tree-like structure, for
+  instance here
+
+  .. code-block:: text
+     :name: after-parsing
+
+             *
+            / \
+           1   +
+              / \
+             2   2
+
+* notation :gdef:`interpretation`: each syntactic element gets translated to a
+  term, for instance :n:`1` can be interpreted as the natural number
+  :n:`S O` then :n:`2` is interpreted as :n:`S (S O)`, then :n:`2 * 2`
+  as :n:`Nat.mul (S (S O)) (S (S O))` and finally our whole term as
+  :n:`Nat.add (S O) (Nat.mul (S (S O)) (S (S O)))`.
+
+Each step is performed independently and there isn't any kind of
+backtracking from any step to a previous one. Then, the result goes
```

Comment:
First sentence is unclear and needs to be reworked.  The negative statement "there isn't any kind of backtracking" isn't nearly as helpful as a positive statement of what Coq does.  The steps are distinct but inherently interleaved: Coq lexes a bit, parses a bit, interprets a bit and then repeats the cycle.

Reply:
Well, not at all, the three phases are absolutely successive and none has any influence on the previous ones. That's what I try to make clear here. Obviously I failed. Any suggestion welcome (I agree negative statements are not ideal).

Reply:
I think perhaps the subtlety here is that while the steps are independent, they are not independent passes over the entire document, right? The interpretation phase of `Reserved Notation` can modify the lexer for subsequent statements.  Perhaps starting with "On each document sentence, each step is performed sequentially and independently." or similar?